src/software/networking/ssl_proto_communication.py

```
@@ -0,0 +1,96 @@
+        self.socket.connect(("", port))
+
+    def send(self, proto):
+        """
+        Send the proto through the socket.
+
+        :param proto proto to send
+        """
+        # https://cwiki.apache.org/confluence/display/GEODE/Delimiting+Protobuf+Messages
+        size = proto.ByteSize()
+
+        # Send a request to the host with the size of the message
+        self.socket.send(encoder._VarintBytes(size) + proto.SerializeToString())
+
+    def receive(self, proto_type):
+        """
+        Receives proto(s) on the socket and returns them, given the proto type to expect. This function is blocking
+
+        :param proto_type the proto type to parse received data as
+
+        :return: a list of protos of the provided type that were received from the socket
+
+        :raises TypeError:                      if the given proto_type isn't a known proto type
+        :raises SslSocketProtoParseException:   if the received data from the socket isn't parseable as the given proto
+        """
+        responses = list()
+
+        try:
+            proto_type()
+        except NameError:
+            raise TypeError(f"Unknown proto type: '{proto_type}'")
+
+        response_data = self.socket.recv(SslSocket.RECEIVE_BUFFER_SIZE)
+
+        offset = 0
+        while offset < len(response_data):
+            msg_len, new_pos = decoder._DecodeVarint32(response_data, offset)
+            offset = new_pos
+            ci_output = proto_type()
+
+            # we didn't receive enough data to parse this message, to get more
+            if offset + msg_len > len(response_data):
+                response_data += self.socket.recv(offset + msg_len - len(response_data))
+
+            try:
+                ci_output.ParseFromString(response_data[offset : offset + msg_len])
+            except protobuf_message.DecodeError as err:
+                raise SslSocketProtoParseException(
+                    "ProtoDecode Error parsing proto: " + err.args
+                )
```

Is there a reason we raise exceptions in this file instead of logging? In some sense raising an exception makes it better better for catching bugs earlier by making the error more apparent, but it can possibly be risky to have in an actual game situation, since we don't want Thunderscope to crash.

---

I've been guarding calls to `receive` in a try-except. Do you think logging is still preferable here?